actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+require "erb"
+require "action_dispatch/http/request"
+require "active_support/actionable_error"
+
+module ActionDispatch
+  class ActionableExceptions # :nodoc:
+    cattr_accessor :endpoint, default: "/rails/actions"
+
+    def initialize(app)
+      @app = app
+    end
+
+    def call(env)
+      request = ActionDispatch::Request.new(env)
+      return @app.call(env) unless actionable_request?(request)
+
+      ActiveSupport::ActionableError.dispatch(request.params["error"], request.params["action"])
+
+      redirect_to request.params["location"]
+    end
+
+    private
+      def actionable_request?(request)
```

Comment:
This has to be checked this way, and not mounted as an ordinary route, because we expect other pre-routing middlewares to raise actionable errors, right?

Reply:
Yes. I tried calling the dispatcher from an internal controller, but eventually, it ended up as this middleware. The pending migrations check is inserted quite early:

https://github.com/rails/rails/blob/7c4457447e1136979e8ed9d822700c1041e9efa6/activerecord/lib/active_record/railtie.rb#L84-L89 

This may even be a pattern for rails plugins inserting a feature checking middleware that raises an actionable error.